src/cls.ts

```diff
@@ -43,78 +49,110 @@ export interface PhantomRootContext {
  */
 export type RootContext = RealRootContext|PhantomRootContext;
 
-export type Namespace = CLS.Namespace;
-export type Func<T> = CLS.Func<T>;
+const asyncHooksAvailable = semver.satisfies(process.version, '>=8');
 
-const useAsyncHooks: boolean = semver.satisfies(process.version, '>=8') &&
-    !!process.env.GCLOUD_TRACE_NEW_CONTEXT;
-
-const cls: typeof CLS =
-    useAsyncHooks ? require('./cls-ah') : require('continuation-local-storage');
-
-const TRACE_NAMESPACE = 'com.google.cloud.trace';
+export interface TraceCLSConfig { mechanism: 'async-listener'|'async-hooks'; }
 
 /**
- * Stack traces are captured when a root span is started. Because the stack
- * trace height varies on the context propagation mechanism, to keep published
- * stack traces uniform we need to remove the top-most frames when using the
- * c-l-s module. Keep track of this number here.
+ * An implementation of continuation-local storage for the Trace Agent.
+ * In addition to the underlying API, there is a guarantee that when an instance
+ * of this class is disabled, all context-manipulation methods will either be
+ * no-ops or pass-throughs.
  */
-export const ROOT_SPAN_STACK_OFFSET = useAsyncHooks ? 0 : 2;
+export class TraceCLS implements CLS<RootContext> {
+  private currentCLS: CLS<RootContext>;
+  private spareCLS: CLS<RootContext>;
+  private readonly logger: Logger;
+  private enabled = false;
+
+  /**
+   * Stack traces are captured when a root span is started. Because the stack
+   * trace height varies on the context propagation mechanism, to keep published
+   * stack traces uniform we need to remove the top-most frames when using the
+   * c-l-s module. Keep track of this number here.
+   */
+  readonly rootSpanStackOffset: number;
+
+  constructor(logger: Logger, config: TraceCLSConfig) {
+    this.logger = logger;
+    const uncorrelated: RootContext = {type: SpanDataType.UNCORRELATED};
+    const untraced: RootContext = {type: SpanDataType.UNTRACED};
+    const useAH = config.mechanism === 'async-hooks' && asyncHooksAvailable;
```

Comment:
I would prefer a throw when `config.mechanism === 'async-hooks'` but `asyncHooksAvailable` is false.